### PR TITLE
QuickNet

### PR DESCRIPTION
**Pull request by jamescook106**
This PR adds QuickNet. QuickNet is a Larq Compute Engine (https://github.com/larq/compute-engine) optimised Binary ResNet15.

### CODE REVIEW

**Comment by jamescook106** [comment]
@AdamHillier @lgeiger I need to add the weight files and plots but could you please review the model?

File: larq_zoo/quicknet.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+            padding="same",
+            use_bias=False,
+            activation="relu",
+        )(x)
+
+        x = tf.keras.layers.BatchNormalization(momentum=0.9, epsilon=1e-5)(x)
+        return x
+
+    # LCE Optimised Fast Block
+    def residualfast_block(
```
Comment (AdamHillier) [inline larq_zoo/quicknet.py]:
```suggestion
    def residual_fast_block(
```

File: larq_zoo/quicknet.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            depth_multiplier=8,
+            strides=2,
+            kernel_initializer="he_normal",
+            padding="same",
+            use_bias=False,
+            activation="relu",
+        )(x)
+
+        x = tf.keras.layers.BatchNormalization(momentum=0.9, epsilon=1e-5)(x)
+        return x
```
Comment (AdamHillier) [inline larq_zoo/quicknet.py]: Probably can just `return tf.keras. ...`

File: larq_zoo/quicknet.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+            15: ([2, 3, 4, 4], [64, 128, 256, 512]),
+        }
+        try:
+            return spec[self.num_layers]
+        except Exception:
+            raise ValueError(f"Only specs for layers {list(self.spec.keys())} defined.")
+
+    def LCEFirstLayer(self, x: tf.Tensor) -> tf.Tensor:
+        x = tf.keras.layers.Conv2D(
+            int(self.initial_filters / 8),
```
Comment (AdamHillier) [inline larq_zoo/quicknet.py]:
```suggestion
            self.initial_filters // 8,
```

File: larq_zoo/quicknet.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    num_layers: int = Field(15)
+    initial_filters: int = Field(64)
+
+    input_quantizer = Field(lambda: lq.quantizers.SteSign(clip_value=1.25))
+    kernel_quantizer = Field(lambda: lq.quantizers.SteSign(clip_value=1.25))
+    kernel_constraint = Field(lambda: lq.constraints.WeightClip(clip_value=1.25))
+
+    @property
+    def spec(self):
+        spec = {
```
Comment (AdamHillier) [inline larq_zoo/quicknet.py]: Do we plan to add further specs?

File: larq_zoo/quicknet.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        x = tf.keras.layers.BatchNormalization(momentum=0.9, epsilon=1e-5)(x)
+        if filters != infilters:
+            return tf.keras.layers.concatenate(
+                [residual, tf.keras.layers.add([x, residual])]
+            )
+        else:
+            return tf.keras.layers.add([x, residual])
+
+    def build(self) -> tf.keras.models.Model:
+
```
Comment (AdamHillier) [inline larq_zoo/quicknet.py]:
```suggestion
```

File: larq_zoo/quicknet.py (hunk 4 above)
Comment (jamescook106) [reply larq_zoo/quicknet.py]: I do not add to plan anymore as the large model uses SE

File: larq_zoo/quicknet.py (hunk 2 above)
Comment (jamescook106) [reply larq_zoo/quicknet.py]: I'll update this now

**Comment by jamescook106** [comment]
@lgeiger @AdamHillier this is ready for a proper review

File: larq_zoo/quicknet.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+import tensorflow as tf
+from zookeeper import Field, factory
+
+from larq_zoo import utils
+from larq_zoo.model_factory import ModelFactory
+
+
+@factory
+class QuickNetFactory(ModelFactory):
+
```
Comment (AdamHillier) [inline larq_zoo/quicknet.py]: Could you add a doc string here?

File: larq_zoo/quicknet.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+            input_quantizer=self.input_quantizer,
+            kernel_quantizer=self.kernel_quantizer,
+            kernel_constraint=self.kernel_constraint,
+            kernel_initializer="glorot_normal",
+            use_bias=False,
+            activation="relu",
+            metrics=[],
+        )(x)
+        x = tf.keras.layers.BatchNormalization(momentum=0.9, epsilon=1e-5)(x)
+        if filters != infilters:
```
Comment (lgeiger) [inline larq_zoo/quicknet.py]:
```suggestion
        if downsample:
```

File: larq_zoo/quicknet.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+class QuickNetFactory(ModelFactory):
+
+    num_layers: int = Field(15)
+    initial_filters: int = Field(64)
+
+    input_quantizer = Field(lambda: lq.quantizers.SteSign(clip_value=1.25))
+    kernel_quantizer = Field(lambda: lq.quantizers.SteSign(clip_value=1.25))
+    kernel_constraint = Field(lambda: lq.constraints.WeightClip(clip_value=1.25))
+
+    @property
```
Comment (AdamHillier) [inline larq_zoo/quicknet.py]:
```suggestion
    @Field```

File: larq_zoo/quicknet.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    def spec(self):
```
Comment (AdamHillier) [inline larq_zoo/quicknet.py]:
```suggestion
    def spec(self) -> Tuple[Sequence[int], Sequence[int]]:
```

File: larq_zoo/quicknet.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
+from typing import Optional, Sequence
```
Comment (AdamHillier) [inline larq_zoo/quicknet.py]:
```suggestion
from typing import Optional, Sequence, Tuple
```

**Pull request merged**